controllers/storageoscluster/metrics-exporter_operand.go

```diff
@@ -0,0 +1,153 @@
+	fs              filesys.FileSystem
+	kubectlClient   kubectl.KubectlClient
+
+	prometheusCRDsInstalled bool
+	prometheusCRDsLock      chan struct{}
+}
+
+var _ operand.Operand = &MetricsExporterOperand{}
+
+func (c *MetricsExporterOperand) Name() string                                           { return c.name }
+func (c *MetricsExporterOperand) Requires() []string                                     { return c.requires }
+func (c *MetricsExporterOperand) RequeueStrategy() operand.RequeueStrategy               { return c.requeueStrategy }
+func (c *MetricsExporterOperand) PostReady(ctx context.Context, obj client.Object) error { return nil }
+
+// dependenciesFound dictates wether the dependencies are present
+// and thus we can apply the metrics-exporter resources
+// currently we depend on Prometheus CRDs
+func (c *MetricsExporterOperand) dependenciesFound() bool {
+	return c.prometheusCRDsInstalled
+}
+
+func (c *MetricsExporterOperand) updateDependenciesState(currentState bool) {
+	c.prometheusCRDsLock <- struct{}{}
+	defer func() {
+		<-c.prometheusCRDsLock
+	}()
+
+	c.prometheusCRDsInstalled = currentState
+}
+
+func (c *MetricsExporterOperand) ReadyCheck(ctx context.Context, obj client.Object) (bool, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "MetricsExporterOperand.ReadyCheck")
+	defer span.End()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	if !c.dependenciesFound() {
+		return false, nil
+	}
+
+	nodeDS := &appsv1.DaemonSet{}
+	key := client.ObjectKey{Name: "storageos-node", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, nodeDS); err != nil {
+		return false, err
+	}
+
+	metricsDS := &appsv1.DaemonSet{}
+	key = client.ObjectKey{Name: "storageos-metrics-exporter", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, metricsDS); err != nil {
+		return false, err
+	}
+
+	// metrics-exporter only returns complete metrics when there's one instance running
+	// alongside each node container. Failing to do so means the metrics can be missing
+	// data for some volumes.
+	if metricsDS.Status.NumberReady == nodeDS.Status.DesiredNumberScheduled {
+		log.Info("metrics-exporter is ready", "NumberReady", metricsDS.Status.NumberReady)
+		return true, nil
+	}
+
+	log.Info("metrics-exporter not ready")
+	return false, nil
+}
+
+func (c *MetricsExporterOperand) Ensure(ctx context.Context, obj client.Object, ownerRef metav1.OwnerReference) (eventv1.ReconcilerEvent, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "MetricsExporterOperand.Ensure")
+	defer span.End()
+
+	b, err := declarative.NewBuilder(metricsExporterPackage, c.fs, declarative.WithKubectlClient(c.kubectlClient))
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+
+	err = b.Apply(ctx)
+	// the error type is lost along the way
+	if err != nil && strings.Contains(err.Error(), "no matches for kind \"ServiceMonitor\" in version \"monitoring.coreos.com/v1\"") {
```

Comment:
Would be nice to check CRD instead of this string thing `k get crd | grep prometheus`

Reply:
Agree, but I couldn't fetch the CRD with the current client.Get we have available to use here
What do you suggest?

Reply:
I haven't tried, just this looks promising: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apimachinery/pkg/api/meta/errors.go#L111. Could you please give a ride or find an other solution for that specific error?

Reply:
I tried all the errors and methods on that package plus a combination of type assertion, errors.Is and errors.As
None of those things worked here. Looking through the call stack to see why, at some point they use a wrapper package for errors that transforms the error stack on an array of errors as their string representations (as far as I could understand it)

Reply:
Ok thanks :( ... Go vs. error handling 0:1